### ISSUE TITLE
The lenet.cpp example runs ok ,but accuracy does not improve

### ISSUE THREAD

**Issue by zhubuntu** [dmlc/MXNet.cpp#2]
I'm trying to use the lenet.cpp example.It can runs,But the accuracy always remain 0.0980952.
![qq 20160223102919](https://cloud.githubusercontent.com/assets/15666683/13240310/9746a54e-da1c-11e5-91ca-c61107d33ae9.jpg)

I did make some changes about the lenet.cpp
from 

``` cpp
    Symbol conv1_w("conv1_w"), conv1_b("conv1_b");
    Symbol conv2_w("conv2_w"), conv2_b("conv2_b");
    Symbol conv3_w("conv3_w"), conv3_b("conv3_b");
    Symbol fc1_w("fc1_w"), fc1_b("fc1_b");
    Symbol fc2_w("fc2_w"), fc2_b("fc2_b");
```

to

```
    Symbol conv1_w = Symbol::Variable("conv1_w");
    Symbol conv1_b = Symbol::Variable("conv1_b");
    Symbol conv2_w = Symbol::Variable("conv2_w");
    Symbol conv2_b = Symbol::Variable("conv2_b");
    Symbol conv3_w = Symbol::Variable("conv3_w");
    Symbol conv3_b = Symbol::Variable("conv3_b");
    Symbol fc1_w = Symbol::Variable("fc1_w");
    Symbol fc1_b = Symbol::Variable("fc1_b");
    Symbol fc2_w = Symbol::Variable("fc2_w");
    Symbol fc2_b = Symbol::Variable("fc2_b");
```

Because I found It will report error with the former code
`Microsoft C++ 异常: std::length_error，位于内存位置 0x0000009D1AAFE3F0 处。`

And I also refer the  [https://github.com/hjk41/MxNet.cpp/issues/12](https://github.com/hjk41/MxNet.cpp/issues/12),change the code
from 
`args_map["data"] =
      NDArray(Shape(batch_size, 1, W, H), ctx_dev, false);`
to

```
args_map["data"] =
      NDArray(Shape(batch_size, 1, W, H), ctx_dev, false);
args_map["data_label"] =
        NDArray(Shape(batch_size), ctx_dev, false);
```

**Comment by zhangchen-qinyinghua** [comment]
I am not familiar with Windows programming, however, Symbol::Variable(str) and Symbol(str) should be in some degree the same.

As for the accuracy issue, could you please paste me your full code here?
In fact, the change you refer the https://github.com/hjk41/MxNet.cpp/issues/12 should not be necessary.

**Comment by hjk41** [comment]
@zhubuntu are you using cpu or gpu? If you are using GPU on windows, then it might be related to this bug: https://github.com/dmlc/mxnet/issues/1228 . In this case, try the latest binary build to see if it fixes your problem.

**Comment by zhubuntu** [comment]
@zhangchen-qinyinghua  @hjk41 I try the latest binary and C++ wrapper, it works fine now.But there are some questions.
- The accuracy is not very good.The old code  written by @zhangchen-qinyinghua before using this C++ wrapper to train mnist can achieve 0.965867 at the first epoch.But after use this C++ Wrapper, It can only achieve 0.869683 at the 20 epoch.After I delete the conv3,thnh3 and pool3 layer,the old and new code create the same net structure. But the new code still can only achieve 0.93333 at the 10  epoch.
- The `Symbol::Variable(str)` and `Symbol(str)` problem stll exit.
- I add the`#include <stdint.h>` to lenet.cpp,because the error information

`mxnet.cpp-master\include\c_api.h(986): error C2061: 语法错误: 标识符“uint64_t”`
- I also add the `#include<numeric>` to optimizer.h because the error information

`mxnet.cpp-master\include\optimizer.hpp(42): error C2039: “accumulate”: 不是“std”的成员`
- the program may  crash for
  `0x00007FFD8CEE1F08 处(位于 MxnetTestApp.exe 中)最可能的异常: Microsoft C++ 异常: std::bad_function_call，位于内存位置 0x000000B08C0FFB60 处。
  0x00007FFD54DB5EC4 (libmxnet.dll) (MxnetTestApp.exe 中)处有未经处理的异常:  请求了严重的程序退出。`
  I met this problem for two times,one is at 20 epochs ,another is at 34 epochs.
- This is the full code 

```
#include <iostream>
#include <fstream>
#include <map>
#include <string>
#include <stdint.h>
#include <vector>
#include "MxNetCpp.h"
using namespace std;
using namespace mxnet::cpp;

class Lenet {
public:
    Lenet()
        : ctx_cpu(Context(DeviceType::kCPU, 0)),
        ctx_dev(Context(DeviceType::kGPU, 0)) {}
    void Run() {
        /*
        * LeCun, Yann, Leon Bottou, Yoshua Bengio, and Patrick Haffner.
        * "Gradient-based learning applied to document recognition."
        * Proceedings of the IEEE (1998)
        * */

        /*define the symbolic net*/
        Symbol data = Symbol::Variable("data");
        Symbol data_label = Symbol::Variable("data_label");
        Symbol conv1_w = Symbol::Variable("conv1_w");
        Symbol conv1_b = Symbol::Variable("conv1_b");
        Symbol conv2_w = Symbol::Variable("conv2_w");
        Symbol conv2_b = Symbol::Variable("conv2_b");
        Symbol conv3_w = Symbol::Variable("conv3_w");
        Symbol conv3_b = Symbol::Variable("conv3_b");
        Symbol fc1_w = Symbol::Variable("fc1_w");
        Symbol fc1_b = Symbol::Variable("fc1_b");
        Symbol fc2_w = Symbol::Variable("fc2_w");
        Symbol fc2_b = Symbol::Variable("fc2_b");

        Symbol conv1 =
            Convolution("conv1", data, conv1_w, conv1_b, Shape(5, 5), 20);
        Symbol tanh1 = Activation("tanh1", conv1, ActivationActType::tanh);
        Symbol pool1 = Pooling("pool1", tanh1, Shape(2, 2),
            PoolingPoolType::max, Shape(2, 2));

        Symbol conv2 = Convolution("conv2", pool1, conv2_w, conv2_b,
            Shape(5, 5), 50);
        Symbol tanh2 = Activation("tanh2", conv2, ActivationActType::tanh);
        Symbol pool2 = Pooling("pool2", tanh2, Shape(2, 2),
            PoolingPoolType::max, Shape(2, 2));

        Symbol conv3 = Convolution("conv3", pool2, conv3_w, conv3_b,
            Shape(2, 2), 500);
        Symbol tanh3 = Activation("tanh3", conv3, ActivationActType::tanh);
        Symbol pool3 = Pooling("pool3", tanh3, Shape(2, 2),
            PoolingPoolType::max, Shape(1, 1));

        Symbol flatten = Flatten("flatten", pool3);
        Symbol fc1 = FullyConnected("fc1", flatten, fc1_w, fc1_b, 500);
        Symbol tanh4 = Activation("tanh4", fc1, ActivationActType::tanh);
        Symbol fc2 = FullyConnected("fc2", tanh4, fc2_w, fc2_b, 10);

        Symbol lenet = SoftmaxOutput("softmax", fc2, data_label);

        for (auto s : lenet.ListArguments()) {
            LG << s;
        }

        /*setup basic configs*/
        int val_fold = 1;
        int W = 28;
        int H = 28;
        int batch_size = 42;
        int max_epoch = 100000;
        float learning_rate = 1e-4;
        float weight_decay = 1e-4;

        /*prepare the data*/
        vector<float> data_vec, label_vec;
        size_t data_count = GetData(&data_vec, &label_vec);
        const float *dptr = data_vec.data();
        const float *lptr = label_vec.data();
        NDArray data_array = NDArray(Shape(data_count, 1, W, H), ctx_cpu,
            false);  // store in main memory, and copy to
        // device memory while training
        NDArray label_array =
            NDArray(Shape(data_count), ctx_cpu,
            false);  // it's also ok if just store them all in device memory
        data_array.SyncCopyFromCPU(dptr, data_count * W * H);
        label_array.SyncCopyFromCPU(lptr, data_count);
        data_array.WaitToRead();
        label_array.WaitToRead();

        size_t train_num = data_count * (1 - val_fold / 10.0);
        train_data = data_array.Slice(0, train_num);
        train_label = label_array.Slice(0, train_num);
        val_data = data_array.Slice(train_num, data_count);
        val_label = label_array.Slice(train_num, data_count);

        LG << "here read fin";

        /*init some of the args*/
        // map<string, NDArray> args_map;
        args_map["data"] =
            NDArray(Shape(batch_size, 1, W, H), ctx_dev, false);
        args_map["data"] = data_array.Slice(0, batch_size).Copy(ctx_dev);
        args_map["data_label"] = label_array.Slice(0, batch_size).Copy(ctx_dev);
        NDArray::WaitAll();

        LG << "here slice fin";
        /*
        * we can also feed in some of the args other than the input all by
        * ourselves,
        * fc2-w , fc1-b for example:
        * */
        // args_map["fc2_w"] =
        // NDArray(mshadow::Shape2(500, 4 * 4 * 50), ctx_dev, false);
        // NDArray::SampleGaussian(0, 1, &args_map["fc2_w"]);
        // args_map["fc1_b"] = NDArray(mshadow::Shape1(10), ctx_dev, false);
        // args_map["fc1_b"] = 0;

        lenet.InferArgsMap(ctx_dev, &args_map, args_map);
        Optimizer opt("ccsgd", learning_rate, weight_decay);
        opt.SetParam("momentum", 0.9)
            .SetParam("rescale_grad", 1.0)
            .SetParam("clip_gradient", 10);

        for (int ITER = 0; ITER < max_epoch; ++ITER) {
            size_t start_index = 0;
            while (start_index < train_num) {
                if (start_index + batch_size > train_num) {
                    start_index = train_num - batch_size;
                }
                args_map["data"] =
                    train_data.Slice(start_index, start_index + batch_size)
                    .Copy(ctx_dev);
                args_map["data_label"] =
                    train_label.Slice(start_index, start_index + batch_size)
                    .Copy(ctx_dev);
                start_index += batch_size;
                NDArray::WaitAll();

                Executor *exe = lenet.SimpleBind(ctx_dev, args_map);
                exe->Forward(true);
                exe->Backward();
                exe->UpdateAll(&opt, learning_rate, weight_decay);

                delete exe;
            }

            LG << "Iter " << ITER
                << ", accuracy: " << ValAccuracy(batch_size * 10, lenet);
        }
    }

private:
    Context ctx_cpu;
    Context ctx_dev;
    map<string, NDArray> args_map;
    NDArray train_data;
    NDArray train_label;
    NDArray val_data;
    NDArray val_label;

    size_t GetData(vector<float> *data, vector<float> *label) {
        const char *train_data_path = "./train.csv";
        ifstream inf(train_data_path);
        string line;
        inf >> line;  // ignore the header
        size_t _N = 0;
        while (inf >> line) {
            for (auto &c : line) c = (c == ',') ? ' ' : c;
            stringstream ss;
            ss << line;
            float _data;
            ss >> _data;
            label->push_back(_data);
            while (ss >> _data) data->push_back(_data / 256.0);
            _N++;
        }
        inf.close();
        return _N;
    }

    float ValAccuracy(int batch_size, Symbol lenet) {
        size_t val_num = val_data.GetShape()[0];

        size_t correct_count = 0;
        size_t all_count = 0;

        size_t start_index = 0;
        while (start_index < val_num) {
            if (start_index + batch_size > val_num) {
                start_index = val_num - batch_size;
            }
            args_map["data"] =
                val_data.Slice(start_index, start_index + batch_size).Copy(ctx_dev);
            args_map["data_label"] =
                val_label.Slice(start_index, start_index + batch_size).Copy(ctx_dev);
            start_index += batch_size;
            NDArray::WaitAll();

            Executor *exe = lenet.SimpleBind(ctx_dev, args_map);
            exe->Forward(false);

            const auto &out = exe->outputs;
            NDArray out_cpu = out[0].Copy(ctx_cpu);
            NDArray label_cpu =
                val_label.Slice(start_index - batch_size, start_index).Copy(ctx_cpu);

            NDArray::WaitAll();

            const mx_float *dptr_out = out_cpu.GetData();
            const mx_float *dptr_label = label_cpu.GetData();
            for (int i = 0; i < batch_size; ++i) {
                float label = dptr_label[i];
                int cat_num = out_cpu.GetShape()[1];
                float p_label = 0, max_p = dptr_out[i * cat_num];
                for (int j = 0; j < cat_num; ++j) {
                    float p = dptr_out[i * cat_num + j];
                    if (max_p < p) {
                        p_label = j;
                        max_p = p;
                    }
                }
                if (label == p_label) correct_count++;
            }
            all_count += batch_size;

            delete exe;
        }
        return correct_count * 1.0 / all_count;
    }
};

int main(int argc, char const *argv[]) {
    Lenet lenet;
    lenet.Run();
    return 0;
}
```

**Comment by zhangchen-qinyinghua** [comment]
1.
It should be that the accuracy differs because of the different training process and the usage of Optimizer.
The weight decay and decreasing learning rate must slow down the learning, I suppose.
I get the following result:

```
[14:45:27] lenet.cpp:146: Iter 223, accuracy: 0.952381
```

It's no good of course, data augmentation with drop out and bigger net should take the accuracy exceeding 99%

2.
As for the link and compile error, maybe it's because of the different compiler and platform?
If so, it will be good if you can make a PR to fix it.

**Comment by hjk41** [comment]
For the compiler errors, I think we should add #include <stdint.h> to
c_api.h directly. You can make a PR to mxnet for this.

On Thu, Feb 25, 2016 at 6:50 PM, Zhang Chen notifications@github.com
wrote:

> 1.
> It should be that the accuracy differs because of the different training
> process and the usage of Optimizer.
> The weight decay and decreasing learning rate must slow down the learning,
> I suppose.
> I get the following result:
> 
> [14:45:27] lenet.cpp:146: Iter 223, accuracy: 0.952381
> 
> It's no good of course, data augmentation with drop out and bigger net
> should take the accuracy exceeding 99%
> 
> 2.
> As for the link and compile error, maybe it's because of the different
> compiler and platform?
> If so, it will be good if you can make a PR to fix it.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/dmlc/MXNet.cpp/issues/2#issuecomment-188724258.

## 

HONG Chuntao
System Research Group
Microsoft Research Asia

**Comment by zhubuntu** [comment]
@zhangchen-qinyinghua I compare the  two codes.All the parameters are same.
So I don't think this is  because the  weight decay and decreasing learning rate.I'll try to find why.

**Comment by thirdwing** [comment]
@zhubuntu I am facing the same problem. Have you solved this?
